File: (참고용)collectors/core/backend_adapter.js
/**
 * Backend Adapter for Framework Integration (Node.js)
 * Framework 프로젝트의 backend API와 통신하는 어댑터 (Node.js 버전)
 * 
 * Python 수집기를 Node.js에서 호출하거나 직접 사용할 수 있습니다.
 */

import { exec } from 'child_process';
import { promisify } from 'util';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const execAsync = promisify(exec);
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export class BackendAdapter {
  constructor(backendUrl = null) {
    this.backendUrl = backendUrl || process.env.BACKEND_API_URL || 'http://localhost:3000';
    this.apiEndpoint = `${this.backendUrl}/api/collectors/raw`;
  }

  /**
   * Python 수집기 실행
   * @param {string} collectorType - 수집기 타입 (예: 'sgis', 'sbiz')
   * @param {object} params - 실행 파라미터
   * @returns {Promise<object>} 실행 결과
   */
  async executePythonCollector(collectorType, params = {}) {
    try {
      const collectorPath = join(__dirname, '..', 'core', 'runner.py');
      const paramsJson = JSON.stringify(params);
      
      const { stdout, stderr } = await execAsync(
        `python -m core.runner run ${collectorType} --params '${paramsJson}'`,
        { cwd: join(__dirname, '..') }
      );

      if (stderr) {
        console.warn(`Python collector stderr: ${stderr}`);
      }

      return JSON.parse(stdout || '{}');
    } catch (error) {
      console.error(`Failed to execute Python collector ${collectorType}:`, error);
      throw error;
    }
  }

  /**
   * 수집된 데이터를 backend API로 전송
   * @param {string} collectorType - 수집기 타입
   * @param {any} rawData - 원시 데이터
   * @param {object} metadata - 메타데이터
   * @returns {Promise<object>} API 응답
   */
  async sendRawData(collectorType, rawData, metadata = {}) {
    try {
      const payload = {
        collector_type: collectorType,
        data: rawData,
        timestamp: new Date().toISOString(),
        metadata: metadata
      };

      console.log(`Sending data to backend: ${collectorType}`, {
        endpoint: this.apiEndpoint,
        dataSize: JSON.stringify(rawData).length
      });

      const response = await fetch(this.apiEndpoint, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(payload)
      });

      if (!response.ok) {
        throw new Error(`Backend API error: ${response.status} ${response.statusText}`);
      }

      const result = await response.json();
      console.log(`Successfully sent data to backend: ${collectorType}`, {
        status: result.status,
        message: result.message
      });

      return result;
    } catch (error) {
      console.error(`Failed to send data to backend: ${collectorType}`, {
        error: error.message,
        endpoint: this.apiEndpoint
      });
      throw error;
    }
  }

  /**
   * Backend API 상태 확인
   * @returns {Promise<object>} 상태 정보
   */
  async checkBackendStatus() {
    try {
      const healthEndpoint = `${this.backendUrl}/api/health`;
      const response = await fetch(healthEndpoint, { timeout: 5000 });
      
      if (!response.ok) {
        throw new Error(`Health check failed: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.warn(`Backend health check failed: ${error.message}`);
      return { status: 'unavailable', error: error.message };
    }
  }

  /**
   * 수집 결과를 backend로 전송
   * @param {string} collectorType - 수집기 타입
   * @param {object} result - 수집 결과
   * @returns {Promise<object>} API 응답
   */
  async sendCollectionResult(collectorType, result) {
    const metadata = {
      records_collected: result.records_collected || 0,
      errors: result.errors || [],
      start_time: result.start_time,
      end_time: result.end_time
    };

    return this.sendRawData(
      collectorType,
      result.data || {},
      metadata
    );
  }
}

// 싱글톤 인스턴스
let _backendAdapter = null;

/**
 * BackendAdapter 싱글톤 인스턴스 반환
 * @returns {BackendAdapter} 인스턴스
 */
export function getBackendAdapter() {
  if (!_backendAdapter) {
    _backendAdapter = new BackendAdapter();
  }
  return _backendAdapter;
}

File: (참고용)collectors/core/backend_adapter.py
"""
Backend Adapter for Framework Integration
Framework 프로젝트의 backend API와 통신하는 어댑터

이 모듈은 collectors 폴더에서 수집한 데이터를 backend API로 전송하는 역할을 합니다.
"""

import os
import json
import requests
from typing import Dict, Any, Optional
from datetime import datetime
import logging

logger = logging.getLogger(__name__)


class BackendAdapter:
    """
    Backend API와 통신하는 어댑터 클래스
    
    수집 완료 후 backend의 Raw Store에 데이터를 전송합니다.
    """
    
    def __init__(self, backend_url: Optional[str] = None):
        """
        BackendAdapter 초기화
        
        Args:
            backend_url: Backend API 기본 URL (기본값: 환경 변수에서 읽기)
        """
        self.backend_url = backend_url or os.getenv(
            'BACKEND_API_URL', 
            'http://localhost:3000'
        )
        self.api_endpoint = f"{self.backend_url}/api/collectors/raw"
        
    def send_raw_data(
        self,
        collector_type: str,
        raw_data: Any,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        수집된 원시 데이터를 backend API로 전송
        
        Args:
            collector_type: 수집기 타입 (예: 'sgis', 'sbiz', 'openup')
            raw_data: 수집된 원시 데이터 (dict, list 등)
            metadata: 추가 메타데이터 (선택)
            
        Returns:
            API 응답 결과
            
        Raises:
            requests.RequestException: API 호출 실패 시
        """
        try:
            payload = {
                'collector_type': collector_type,
                'data': raw_data,
                'timestamp': datetime.now().isoformat(),
                'metadata': metadata or {}
            }
            
            logger.info(f"Sending data to backend: {collector_type}", {
                'endpoint': self.api_endpoint,
                'data_size': len(json.dumps(raw_data)) if isinstance(raw_data, (dict, list)) else 0
            })
            
            response = requests.post(
                self.api_endpoint,
                json=payload,
                headers={'Content-Type': 'application/json'},
                timeout=30
            )
            
            response.raise_for_status()
            
            result = response.json()
            logger.info(f"Successfully sent data to backend: {collector_type}", {
                'status': result.get('status'),
                'message': result.get('message')
            })
            
            return result
            
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to send data to backend: {collector_type}", {
                'error': str(e),
                'endpoint': self.api_endpoint
            })
            raise
    
    def check_backend_status(self) -> Dict[str, Any]:
        """
        Backend API 상태 확인
        
        Returns:
            Backend 상태 정보
        """
        try:
            health_endpoint = f"{self.backend_url}/api/health"
            response = requests.get(health_endpoint, timeout=5)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.warning(f"Backend health check failed: {e}")
            return {'status': 'unavailable', 'error': str(e)}
    
    def send_collection_result(
        self,
        collector_type: str,
        result: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        수집 결과를 backend로 전송
        
        Args:
            collector_type: 수집기 타입
            result: 수집 결과 (records_collected, errors 등 포함)
            
        Returns:
            API 응답 결과
        """
        metadata = {
            'records_collected': result.get('records_collected', 0),
            'errors': result.get('errors', []),
            'start_time': result.get('start_time'),
            'end_time': result.get('end_time')
        }
        
        return self.send_raw_data(
            collector_type=collector_type,
            raw_data=result.get('data', {}),
            metadata=metadata
        )


# 전역 인스턴스 (선택적 사용)
_backend_adapter: Optional[BackendAdapter] = None


def get_backend_adapter() -> BackendAdapter:
    """
    BackendAdapter 싱글톤 인스턴스 반환
    
    Returns:
        BackendAdapter 인스턴스
    """
    global _backend_adapter
    if _backend_adapter is None:
        _backend_adapter = BackendAdapter()
    return _backend_adapter

